src/Nethermind/Nethermind.TxPool/Collections/BlobTxDistinctSortedPool.cs

```diff
@@ -0,0 +1,70 @@
+// SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System.Collections.Generic;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Logging;
+
+namespace Nethermind.TxPool.Collections;
+
+public class BlobTxDistinctSortedPool : TxDistinctSortedPool
+{
+    protected const int MaxNumberOfBlobsInBlock = (int)(Eip4844Constants.MaxBlobGasPerBlock / Eip4844Constants.BlobGasPerBlob);
+
+    public BlobTxDistinctSortedPool(int capacity, IComparer<Transaction> comparer, ILogManager logManager)
+        : base(capacity, comparer, logManager)
+    {
+    }
+
+    protected override IComparer<Transaction> GetReplacementComparer(IComparer<Transaction> comparer)
+        => comparer.GetBlobReplacementComparer();
+
+    public virtual bool TryInsertBlobTx(Transaction fullBlobTx, out Transaction? removed)
+        => base.TryInsert(fullBlobTx.Hash, fullBlobTx, out removed);
+
+    public virtual bool TryGetBlobTx(ValueKeccak hash, out Transaction? fullBlobTx)
+        => base.TryGetValue(hash, out fullBlobTx);
+
+    public virtual IEnumerable<Transaction> GetBlobTransactions()
+    {
+        int pickedBlobs = 0;
+        List<Transaction>? blobTxsToReadd = null;
+
+        while (pickedBlobs < MaxNumberOfBlobsInBlock)
+        {
+            Transaction? bestTx = GetFirsts().Min;
+
+            if (bestTx?.Hash is null || bestTx.BlobVersionedHashes is null)
+            {
+                break;
+            }
+
+            if (pickedBlobs + bestTx.BlobVersionedHashes.Length <= MaxNumberOfBlobsInBlock)
+            {
+                yield return bestTx;
+                pickedBlobs += bestTx.BlobVersionedHashes.Length;
+            }
+
+            if (TryRemove(bestTx.Hash))
+            {
+                blobTxsToReadd ??= new(MaxNumberOfBlobsInBlock);
+                blobTxsToReadd.Add(bestTx!);
+            }
+        }
+
+        if (blobTxsToReadd is not null)
+        {
+            foreach (Transaction blobTx in blobTxsToReadd)
+            {
+                TryInsert(blobTx.Hash, blobTx!, out Transaction? removed);
```

Comment:
Because we are working on main blob tx collection, not on a copy. And we can't just remove tx when we are preparing block - it might not be included, or we might want to improve it (so prepare again)

But I also don't like this, I have another idea for this and will try to refactor it

Reply:
It's removed after refactoring